homeassistant/components/stream/worker.py

```diff
@@ -350,19 +358,23 @@ def is_video(packet: av.Packet) -> Any:
     try:
         if audio_stream and unsupported_audio(container_packets.peek(), audio_stream):
             audio_stream = None
-            container_packets = PeekIterator(
-                filter(dts_validator.is_valid, container.demux(video_stream))
-            )
+            container_packets.replace_underlying_iterator(container.demux(video_stream))
```

Comment:
Is this the same as not having the is valid filter here?

Reply:
Oh no, forgot to put it back in, good catch.

Reply:
What is the replace iterator needed for? I'm not exactly seeing the problem here of just re-opening the iterator and rewinding everything? This seems like it will replay the packets in the buffer twice, once in the buffer and once in the new container. Also the buffer may contain stray audio packets.

I likely am missing something so just spelling out my thought process

Reply:
Also if you feel like this refactor is risky I can revert the PR and figure out how to roll forward with more tests.

Reply:
At least with RTSP streams, creating a new iterator via demux() does not seem to replay the previously played packets:
```
>>> import av
>>> container=av.open("rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mov")
>>> next(container.demux()).dts
0
>>> next(container.demux()).dts
18750
```

Reply:
It doesn't look like stray packets are a problem since we check the source in `mux_packet` anyway. I think this second `demux(video_stream)` here was just to filter out audio packets earlier on in the pipeline, and we don't really need to do it at all.

Reply:
Ok makes sense that it doesn't start over now that I think about it.

Are you saying it's enough to just have the line that sets audio stream to none then leave the container/packets iterator allone?

Reply:
If we leave the iterator alone, I don't think there is a muxing problem - I had a look back and it seems that since https://github.com/home-assistant/core/commit/89aaeb3c351a5f9c3fe28bbd62066b4ab5cd665e there is a check for the correct stream in `mux_packet`, so nothing will get muxed incorrectly and crash the worker.
However, in the case where audio exists but is unsupported, we will waste time looping through those packets when they could have just been removed upstream.

Reply:
Ok thanks for resolving this!